content/docs/specs/remote_write_spec_2_0.md

```diff
@@ -134,7 +134,30 @@ Receivers ingesting all samples successfully MUST return a 200 HTTP status code.
 
 Receivers MUST NOT return a 200 HTTP status code if any of the samples were not written successfully (e.g. on a [partial write](#partial-write) or a full write rejection). In such a case, the Receiver MUST provide a human-readable error message in the response body. The Receiver's error SHOULD contain information about the amount of the samples being rejected and for what reasons. Senders MUST NOT try and interpret the error message and SHOULD log it as is.
 
-The following subsections specify Sender and Receiver semantics around different write error cases.
+The following subsections specify Sender and Receiver semantics around headers and different write error cases.
+
+#### Response Headers
+
+<!---
+Rationales: https://github.com/prometheus/prometheus/issues/14359
+-->
+For all HTTP status codes, Receivers MUST send back the HTTP response `X-Prometheus-Remote-Write-Received-` type of the header indicating write statistics, i.e. how many Samples, Histograms and Exemplars were successfully received by the Receiver. Receivers MAY omit each header, if its value is zero.
+
+Receivers MUST send those response headers, even if they use asynchronous or eventual write mechanisms e.g. when Receivers can't confirm that downstream storage actually written them in durable way. The headers are only to confirm how many Samples, Histograms and Exemplars were accepted by the Receiver, especially useful on [Partial Write](#partial-write) situations.
+
+Each header value MUST be a single 64 bytes integer. The header names MUST be as follows:
+
+```
+X-Prometheus-Remote-Write-Received-Samples <integer; count of all written Samples from this request>
+X-Prometheus-Remote-Write-Received-Histograms <integer; count of all written Histogram samples from this request>
+X-Prometheus-Remote-Write-Received-Exemplars <integer; count of all written Exemplars from this request>
+```
```

Comment:
```suggestion
For all HTTP status codes, Receivers MUST send back the HTTP response `X-Prometheus-Remote-Write-Accepted-` type of the header indicating write statistics, i.e. how many Samples, Histograms and Exemplars were successfully accepted/ingested by the Receiver. Receivers MAY omit each header, if its value is zero.

Receivers MUST send those response headers, even if they use asynchronous or eventual write mechanisms e.g. when Receivers can't confirm that downstream storage actually written them in durable way. The headers are only to confirm how many Samples, Histograms and Exemplars were accepted by the Receiver, especially useful on [Partial Write](#partial-write) situations.

Each header value MUST be a single 64 bytes integer. The header names MUST be as follows:

```
X-Prometheus-Remote-Write-Accepted-Samples <integer; count of all written Samples from this request>
X-Prometheus-Remote-Write-Accepted-Histograms <integer; count of all written Histogram samples from this request>
X-Prometheus-Remote-Write-Accepted-Exemplars <integer; count of all written Exemplars from this request>
```

Reply:
I think we should change `Received` to `Accepted` to make it clearer that the values in the header are supposed to be the # of each data type successfully ingested rather than just the # of each that were received in the request.

Reply:
Hm, good point. We can.

I wonder if naming it "accepted" won't conflict with the statement added to the spec in this PR about eventual storage.

What if there is Receiver by design stores things in some queue and cannot confirm durability?

I guess "accept" does not mean "durablyStoredForverWeDoubleChecked" so it's fine too 🙈

Reply:
That's a whole different can of worms, and I think outside the scope for RW spec. We can really only specify that they should tell us how many of each samples they are going to accept, or reject. The 200 status code or retryable error is only able to tell the Sender "I can accept these" or "some of these are not accepted but retryable". Whether or not that data has been committed to persistent storage is outside RW's concerns IMO. Every RW Receiver implementer is going to have different guarantees about how they accept data and the timeline on which accepted data is committed to persistent storage/available for reads.

Reply:
Let's do accept!